### PR TITLE
feat(c11): Add additional warning flags

### PR DESCRIPTION
**Pull request by miguelafsilva5**
This PR introduces several compiler flags that issue warnings from additional sources.
The goal of these flags is to help complying with some MISRA directives. The flags include: forcing explicit casts, adding missing protypes, removing unused macros and functions, etc.

This PR depends on #146 

The code was tested in the following platforms:
- qemu-aarch64-virt
- qemu-riscv64-virt

### CODE REVIEW

**Comment by DavidMCerdeira** [comment]
@miguelafsilva5 note that one of the checks is failing

**Comment by miguelafsilva5** [comment]
> @miguelafsilva5 note that one of the checks is failing

The code was not formated properly. It is fixed in the recent commits.

**Comment by miguelafsilva5** [comment]
> @miguelafsilva5 note that one of the checks is failing

The code was not formatted properly. It is fixed in the recent commits.

File: Makefile [hunk 1]
```diff
@@ -237,7 +249,7 @@ $(build_dir)/%.d : $(src_dir)/%.[c,S]
 
 $(objs-y):
 	@echo "Compiling source	$(patsubst $(cur_dir)/%, %, $<)"
-	@$(cc) $(CFLAGS) -c $< -o $@
+	-@$(cc) $(CFLAGS) -c $< -o $@
```
Comment (DavidMCerdeira) [inline Makefile:252]: Should remove '-'

File: src/arch/armv8/cache.c [hunk 2]
```diff
@@ -42,6 +42,9 @@ void cache_arch_enumerate(struct cache* dscrp)
                 case CLIDR_CTYPE_UN:
                     dscrp->type[i] = UNIFIED;
                     break;
+                default:
+                    dscrp->type[i] = UNIFIED;
```
Comment (DavidMCerdeira) [inline src/arch/armv8/cache.c:46]: As we hadn't had a default case before, shouldn't this return an error / hang?  or at least a warning

File: src/arch/armv8/inc/arch/gicv3.h [hunk 3]
```diff
@@ -107,6 +107,8 @@ static inline void gich_write_lr(size_t i, uint64_t val)
         case 15:
             sysreg_ich_lr15_el2_write(val);
             break;
+        default:
+            break;
```
Comment (DavidMCerdeira) [inline src/arch/armv8/inc/arch/gicv3.h:111]: Shouldn't this return an error / hang? or at least a warning

File: src/arch/armv8/psci.c [hunk 4]
```diff
@@ -162,7 +165,7 @@ static int32_t psci_affinity_info_handler(unsigned long target_affinity,
 
 static int32_t psci_features_handler(uint32_t feature_id)
 {
-    int32_t ret = PSCI_E_NOT_SUPPORTED;
+    int32_t ret;
```
Comment (DavidMCerdeira) [inline src/arch/armv8/psci.c:168]: just a nit pick: I think it's fine to leave this initialized even if the value is overwritten

File: Makefile (hunk 1 above)
Comment (miguelafsilva5) [reply Makefile]: Good catch. Fixed in commit 83baede9561218058ae4cfd5af469e84fc96ad7c

File: src/arch/armv8/psci.c (hunk 4 above)
Comment (miguelafsilva5) [reply src/arch/armv8/psci.c]: Fixed in commit 9d3994a088888aef92cfe7e357356b385923936b

File: src/arch/armv8/inc/arch/gicv3.h (hunk 3 above)
Comment (josecm) [reply src/arch/armv8/inc/arch/gicv3.h:111]: We've debated this. But is is just maintaining the same behavior as before. Maybe we could "fix" it in another PR.

**Review by josecm — CHANGES_REQUESTED** [review]

File: src/arch/armv8/inc/arch/gic.h [hunk 5]
```diff
@@ -315,10 +315,10 @@ typedef uint64_t gic_lr_t;
 #define GICH_LR_CPUID_LEN     (3)
 
 #define GICH_LR_VID_MSK       BIT_MASK(GICH_LR_VID_OFF, GICH_LR_VID_LEN)
-#define GICH_LR_VID(LR)       (bit_extract(LR, GICH_LR_VID_OFF, GICH_LR_VID_LEN))
+#define GICH_LR_VID(LR)       (bit64_extract(LR, GICH_LR_VID_OFF, GICH_LR_VID_LEN))
 
 #define GICH_LR_CPUID_MSK     BIT_MASK(GICH_LR_CPUID_OFF, GICH_LR_CPUID_LEN)
-#define GICH_LR_CPUID(LR)     (bit_extract(LR, GICH_LR_CPUID_OFF, GICH_LR_CPUID_LEN))
+#define GICH_LR_CPUID(LR)     (bit64_extract(LR, GICH_LR_CPUID_OFF, GICH_LR_CPUID_LEN))
```
[the remark this answers is not in the document]
Comment (miguelafsilva5) [reply src/arch/armv8/inc/arch/gic.h:322]: How do you propose that? Those lines are already inside a ifdef macro. More specifically inside the GICV3 versions.

File: src/arch/armv8/gicv3.c [hunk 6]
```diff
@@ -43,29 +43,29 @@ static inline void gicr_init()
...
 
     for (size_t i = 0; i < GIC_NUM_PRIO_REGS(GIC_CPU_PRIV); i++) {
         state->priv_IPRIORITYR[i] = gicr[cpu()->id].IPRIORITYR[i];
     }
 
-    state->HCR = sysreg_ich_hcr_el2_read();
+    state->HCR = (uint32_t)sysreg_ich_hcr_el2_read();
     for (size_t i = 0; i < gich_num_lrs(); i++) {
-        state->LR[i] = gich_read_lr(i);
+        state->LR[i] = (uint32_t)gich_read_lr(i);
```
[the remark this answers is not in the document]
Comment (miguelafsilva5) [reply src/arch/armv8/gicv3.c]: Fixed in commit 03942696798f6a0da8963533065cc388609333fb . I found other instance where it should have been `gic_lr_t ` and fixed it too.

File: src/arch/armv8/inc/arch/gicv2.h [hunk 7]
```diff
@@ -20,7 +20,7 @@ static inline uint64_t gich_read_lr(size_t i)
 static inline void gich_write_lr(size_t i, uint64_t val)
 {
     if (i < NUM_LRS) {
-        gich->LR[i] = val;
+        gich->LR[i] = (uint32_t)val;
```
[the remark this answers is not in the document]
Comment (miguelafsilva5) [reply src/arch/armv8/inc/arch/gicv2.h]: Fixed in commit 03942696798f6a0da8963533065cc388609333fb . I found other instance where it should have been `gic_lr_t ` and fixed it too.

File: src/arch/armv8/vgic.c [hunk 8]
```diff
@@ -708,12 +711,13 @@ void vgic_emul_generic_access(struct emul_access* acc, struct vgic_reg_handler_i
     size_t field_width = handlers->field_width;
     size_t first_int = (GICD_REG_MASK(acc->addr) - handlers->regroup_base) * 8 / field_width;
     unsigned long val = acc->write ? vcpu_readreg(cpu()->vcpu, acc->reg) : 0;
-    unsigned long mask = (1ull << field_width) - 1;
-    bool valid_access = (GIC_VERSION == GICV2) || !(gicr_access ^ gic_is_priv(first_int));
+    unsigned long mask = (1ul << field_width) - 1;
```
[the remark this answers is not in the document]
Comment (miguelafsilva5) [reply src/arch/armv8/vgic.c]: Fixed in commit 03942696798f6a0da8963533065cc388609333fb . There were a couple of stray `ull` that were not capitalized and now they are.

File: src/arch/armv8/inc/arch/gic.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (josecm) [reply src/arch/armv8/inc/arch/gic.h:322]: No they arent. But we should fix this in another commit.

**Review by DavidMCerdeira — APPROVED** [review]

**Pull request merged**